### ISSUE TITLE
VkAllocationCallbacks application requirements

### ISSUE THREAD

**Issue by philiptaylor** [KhronosGroup/Vulkan-Docs#21]
Section 10.1 (Host Memory) says:

> An allocator indicates an error condition by returning NULL from pfnAllocation or pfnReallocation. If this occurs, the implementation should treat it as a run time error [...]

and:

> pfnAllocation must either return NULL (in case of allocation failure or if size is zero) or a valid pointer [...]

If size is zero, the second statement indicates it's okay (and required) for the application to return NULL, but the first statement indicates the NULL return should always be treated as an error, which is inconsistent.

The "must" requirement to return NULL if size is zero also means that C's malloc() (or posix_memalign etc for alignment) is not an acceptable implementation of pfnAllocation, because malloc(0) may simply act as if the size is non-zero and return a non-NULL pointer (which the caller must later free()). I'm not sure if that's an intentional difference - it seems unnecessary and surprising.

The return value of pfnReallocation with pOriginal != NULL, size == 0 also seems unclear. It "must behave similarly to PFN_vkFreeFunction" but that doesn't have a return value. Is it allowed or required or forbidden to return NULL? (In C, realloc(p, 0) may return NULL but is not required to.)

I think it'd be better to follow C's malloc/realloc/free semantics exactly, given that many applications will probably want to simply call standard malloc/realloc/free-like functions with a bit of tracking code around them.

Also:

> pfnReallocation must alter the size of the allocation pOriginal, either by shrinking or growing it, to accommodate the new size.

If the application uses an arena allocator, it can't shrink or free an allocation, so the most efficient thing it can do when asked to reallocate smaller is to return the original pointer and not modify any internal state at all. That's a perfectly sensible thing to do, but it would clearly violate this requirement, so it seems a bad requirement.

(And if I was being pedantic, a standard realloc() returning an entirely new pointer is not really "alter[ing] the size of allocation pOriginal" either.)

It would probably be better to say something like: "If pOriginal is not NULL, and size is not zero, then pfnReallocation must return a valid pointer to a memory allocation of at least size bytes, which may be the same pointer as pOriginal or may be a different pointer". Don't try to prescribe *how* the function chooses such a pointer.

Similar for the "must be freed by the function" in pfnFree. If the application wants to implement pfnFree as a noop, and just free an entire scope's allocations in one go later on, it sounds like it'd be violating that requirement.

**Comment by oddhack** [comment]
Taking the internally for resolution.

N.b. I hope you don't think I'm blowing you off with these comments. The way Khronos works is not intrinsically open, and while we want to be as open as possible and interact effectively with folks on Github, at the moment our process is to move most substantive discussions inside the group and discuss/vote on resolutions that might have IP implications or affect people's drivers. We will try to report back and ask clarifying questions, but for the time being that's what we're trying. Your input is very much appreciated and I think you're hitting on significant areas needing clarification, so please keep it coming.

**Comment by michaelworcester** [comment]
We're discussing this internally as @oddhack said, but I'll try to clarify some points here too.

> I think it'd be better to follow C's malloc/realloc/free semantics exactly, given that many applications will probably want to simply call standard malloc/realloc/free-like functions with a bit of tracking code around them.

This was the original intent; I think the spec language needs to be updated and this is actively being looked at.

> If the application uses an arena allocator, it can't shrink or free an allocation, so the most efficient thing it can do when asked to reallocate smaller is to return the original pointer and not modify any internal state at all. That's a perfectly sensible thing to do, but it would clearly violate this requirement, so it seems a bad requirement.

I don't disagree that this is a bad requirement, but this is also leading down a path where applications are trying to use the allocators for performance. Please note in the spec this is explicitly called out:

> Since most memory allocations are off the critical path, this is not meant as a performance feature. Rather, this can be useful for certain embedded systems, for debugging purposes (e.g. putting a guard page after all host allocations), or for memory allocation logging.

That said I agree the language should be updated to indicate only that the returned pointer should be large enough for the requested allocation size - again, this is actively being discussed.

**Comment by ratchetfreak** [comment]
You can't really just sub in (aligned)malloc, realloc and free to the VkAllocationCallbacks function pointers anyway because the extra void* pUserData parameter they need to take.

So at the very least you have to do a quick forward. If you have that then I don't see a reason why the test for 0 on the parameters would then be a big deal.

```C
void* myMalloc(void* userParam, size_t size,  size_t aligned, VkSystemAllocationScope scope){
     if(size==0)return null;

     return alignedMalloc(size, aligned);
}

void* myRealloc(void* userParam, void* orig, size_t size,  size_t aligned, VkSystemAllocationScope scope){
    if(size == 0) {
        myFree(userParam, orig);
        return null;
    }

    return realloc(orig, size);
}

void myFree(void* userParam, void* mem){
    free(mem);
}
```

Besides that I don't see what the expected behavior is if you allocate with a  `VkAllocationCallbacks` struct but then destroy without one (or vice versa). I expect undefined behavior like you would get with calling free with a pointer that wasn't malloced.

**Comment by philiptaylor** [comment]
@oddhack: That's fine, and thanks for the explanation!

@michaelworcester:
> this is also leading down a path where applications are trying to use the allocators for performance

Okay, makes sense that that shouldn't be encouraged. Maybe a better example is debugging with Valgrind or _CrtSetDbgFlag(_CRTDBG_DELAY_FREE_MEM_DF) where free() doesn't *really* free any memory, it just puts it in some special mode to help detect use-after-free bugs. One could argue over whether that somehow matches what the spec means by "free", but I think it'd be easier if the spec just didn't try to mandate application behaviour beyond what's observable by the implementation (i.e. that it provides a new pointer with at least the requested size, and that the old pointer doesn't have to be accessible any more).

@ratchetfreak: Yeah, it's not hard to work around once you're aware of the issue. I'm just (very mildly) concerned that it's a slightly non-obvious issue, and application developers could easily assume the standard C allocation semantics and write buggy code, and there's seemingly no good reason to make their lives harder in that way. (Of course I can't object to their lives being made harder when there *is* a good reason, because that's a fundamental philosophy behind the whole of Vulkan :-) )

**Comment by ShabbyX** [comment]
Just wanted to pitch in and say being able to use `valgrind` on a vulkan application can be very useful. If `malloc()`/`free()` could be made usable, that would help find a lot of memory leaks and bugs more easily.

**Comment by ratchetfreak** [comment]
Or add the note that implementations shouldnot: use the callbacks in that way. 

As in only call allocate and reallocate with non-zero size and use free to release the memory.

**Comment by philiptaylor** [comment]
> application developers could easily assume the standard C allocation semantics and write buggy code

For example myalloc in https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/sdk-1.0.3/demos/tri.c#L1702 (though myrealloc is even worse since it ignores alignment entirely...)

**Comment by oddhack** [comment]
Should have been fixed in the 1.0.10 spec update.
